conwayste/src/ui/button.rs

```diff
@@ -112,15 +123,46 @@ impl Button {
         let mut b = Button {
             id: None,
             z_index: std::usize::MAX,
-            label: label,
+            label,
             button_color: color_with_alpha(css::DARKCYAN, 0.8),
             draw_mode: DrawMode::fill(),
-            dimensions: dimensions,
+            dimensions,
             hover: false,
+            focused: false,
             borderless: false,
-            action: action,
+            handler_data: context::HandlerData::new(),
         };
         b.center_label_text();
+
+        // setup handler to allow changing appearance when it has keyboard focus
+        let focus_chg = |obj: &mut dyn EmitEvent, _uictx: &mut UIContext, event: &Event| -> Result<Handled, Box<dyn Error>> {
+            let button = obj.downcast_mut::<Button>().unwrap(); // unwrap OK because this will always be Button
+            match event.what {
+                EventType::GainFocus => button.focused = true,
+                EventType::LoseFocus => button.focused = false,
+                _ => unimplemented!("this handler is only for gaining/losing focus"),
+            };
+            Ok(Handled::NotHandled) // allow other handlers for this event type to be activated
+        };
+        b.on(EventType::GainFocus, Box::new(focus_chg.clone())).unwrap(); // unwrap OK b/c not being called within handler
+        b.on(EventType::LoseFocus, Box::new(focus_chg)).unwrap(); // unwrap OK b/c not being called within handler
+
+        // setup handler to forward a space keyboard event to the click handler
+        let keypress = |obj: &mut dyn EmitEvent, uictx: &mut UIContext, event: &Event| -> Result<Handled, Box<dyn Error>> {
```

Comment:
Any reason why this handler (and `focus_chg`) are defined as closures unlike `mouse_move_handler()`? Checkbox seems to go the latter route for all of its handler definitions.

Reply:
I started doing it this way just to show that it could be done. BTW a closure is the only nice way to capture shared state that can't be stored in the widget for whatever reason.

I can fix it.